File: 128.longest-consecutive-sequence.py
#
# @lc app=leetcode id=128 lang=python3
#
# [128] Longest Consecutive Sequence
#

# @lc code=start
class UnionFind:
    def __init__(self, n):
        self.root = list(range(n+1))

    def find(self, i):
        # recursively find root
        if i != self.root[i]:
            self.root[i] = self.find(self.root[i])
        return self.root[i]

    def union(self, i, j): # return operation successful or not
        ri, rj = self.find(i), self.find(j)
        if ri != rj:
            self.root[rj] = ri
    def getParents(self):
        return self.root

class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        if not nums:
            return 0
        # eliminate duplicate number
        newNums = []
        s = set()
        for n in nums:
            if n in s:
                continue
            else:
                s.add(n)
                newNums.append(n)
        # use union find
        # create a map from num of idx
        numIdxMap = {}
        for i, n in enumerate(newNums):
            numIdxMap[n] = i
        uf = UnionFind(len(newNums))
        for i, n in enumerate(newNums):
            if n-1 in numIdxMap:
                uf.union(i, numIdxMap[n-1])
            if n+1 in numIdxMap:
                uf.union(i, numIdxMap[n+1])
        # ??
        for n in range(len(newNums)):
            uf.find(n)
        parents = uf.getParents()
        # count freq of each parent and return the max freq
        d = {}
        for p in parents:
            if d.get(p):
                d[p]+=1
            else:
                d[p]=1
        return max(d.values())
        
# @lc code=end

File: 128.longest-consecutive-sequence.go
/*
 * @lc app=leetcode id=128 lang=golang
 *
 * [128] Longest Consecutive Sequence
 */

// @lc code=start
package main

/*
First, we remove duplicate from nums
Second, we use hashmap to store mapping from number to index
This requires O(n) time to build this map.
Next, for each num in nums (linear traverse),
we do BFS to find the longest sequence this number link to using the hashmap we build
(we need visited array to store whether the number is visted or not to avoid duplication)
*/
func longestConsecutive(nums []int) int {
	newNums := []int{}
	seen := make(map[int]bool)
	for _, n := range nums {
		if seen[n] {
			continue
		}
		seen[n] = true
		newNums = append(newNums, n)
	}

	num2Idx := make(map[int]int)
	for i, n := range newNums {
		num2Idx[n] = i
	}

	visited := make([]bool, len(newNums))
	ans := 0
	for i, n := range newNums {
		if visited[i] {
			continue
		}
		// BFS
		seq := 0
		visited[i] = true
		q := []int{n}
		for len(q) > 0 {
			// pop
			val := q[0]
			q = q[1:]
			seq++
			// neighbors
			if neighborIdx, ok := num2Idx[val+1]; ok {
				if !visited[neighborIdx] {
					visited[neighborIdx] = true
					q = append(q, newNums[neighborIdx])
				}
			}
			if neighborIdx, ok := num2Idx[val-1]; ok {
				if !visited[neighborIdx] {
					visited[neighborIdx] = true
					q = append(q, newNums[neighborIdx])
				}
			}
		}
		if seq > ans {
			ans = seq
		}
	}
	return ans
}

// @lc code=end
